    <h1 id="firstHeading" class="firstHeading">std::domain_error</h1>            <table class="t-dcl-begin"> <tr class="t-dsc-header"> <th> Defined in header <code><a href="../header/stdexcept" title="cpp/header/stdexcept">&lt;stdexcept&gt;</a></code> </th> <th> </th> <th> </th> </tr> <tr class="t-dcl"> <td class="t-dcl-nopad"> <pre data-language="cpp">class domain_error;</pre>
</td> <td class="t-dcl-nopad"> </td> <td class="t-dcl-nopad"> </td> </tr>  </table> <p>Defines a type of object to be thrown as exception. It may be used by the implementation to report domain errors, that is, situations where the inputs are outside of the domain on which an operation is defined.</p>
<p>The standard library components do not throw this exception (mathematical functions report domain errors as specified in <code><a href="../numeric/math/math_errhandling" title="cpp/numeric/math/math errhandling">math_errhandling</a></code>). Third-party libraries, however, use this. For example, <a rel="nofollow" class="external text" href="http://www.boost.org/doc/libs/1_55_0/libs/math/doc/html/math_toolkit/error_handling.html">boost.math</a> throws <code>std::domain_error</code> if <code>boost::math::policies::throw_on_error</code> is enabled (the default setting).</p>
<div class="t-inheritance-diagram"> <div class="center"><div class="noresize">
<map name="ImageMap_1_81497806"><area href="/w/cpp/error/exception" shape="rect" coords="5,4,107,36" alt="cpp/error/exception" title="cpp/error/exception">
<area href="/w/cpp/error/logic_error" shape="rect" coords="135,4,238,36" alt="cpp/error/logic error" title="cpp/error/logic error"></map><img alt="std-domain error-inheritance.svg" src="" width="374" height="40" usemap="#ImageMap_1_81497806">
</div></div> <div style="text-align:center;"> <p>Inheritance diagram.</p>
</div> </div>  <h3 id="Member_functions"> Member functions</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><span class="t-lines"><span>(constructor)</span></span></div> </td> <td> constructs a new <code>domain_error</code> object with the given message <br> <span class="t-mark">(public member function)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><span class="t-lines"><span>operator=</span></span></div> </td> <td> replaces the <code>domain_error</code> object <br> <span class="t-mark">(public member function)</span> </td>
</tr> <tr class="t-dsc"> <td> <div><span class="t-lines"><span>what</span></span></div> </td> <td> returns the explanatory string <br> <span class="t-mark">(public member function)</span> </td>
</tr> </table> <div class="t-member"> <h2 id="std::domain_error::domain_error">  std::domain_error::domain_error </h2> <table class="t-dcl-begin"> <tr class="t-dcl"> <td> <pre data-language="cpp">domain_error( const std::string&amp; what_arg );</pre>
</td> <td> (1) </td> <td class="t-dcl-nopad"> </td> </tr> <tr class="t-dcl t-since-cxx11"> <td> <pre data-language="cpp">domain_error( const char* what_arg );</pre>
</td> <td> (2) </td> <td> <span class="t-mark-rev t-since-cxx11">(since C++11)</span> </td> </tr>  <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(3)</td> <td></td> </tr> <tr class="t-dcl t-until-cxx11"> <td> <pre data-language="cpp">domain_error( const domain_error&amp; other );</pre>
</td>  <td> <span class="t-mark-rev t-until-cxx11">(until C++11)</span> </td> </tr> <tr class="t-dcl t-since-cxx11"> <td> <pre data-language="cpp">domain_error( const domain_error&amp; other ) noexcept;</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx11">(since C++11)</span> </td> </tr>   </table> <div class="t-li1">
<span class="t-li">1-2)</span> Constructs the exception object with <code>what_arg</code> as explanatory string that can be accessed through <a href="exception/what" title="cpp/error/exception/what"><code>what()</code></a>.</div> <div class="t-li1">
<span class="t-li">3)</span> Copy constructor. <span class="t-rev-inl t-since-cxx11"><span>If <code>*this</code> and <code>other</code> both have dynamic type <code>std::domain_error</code> then <code><a href="http://en.cppreference.com/w/cpp/string/byte/strcmp"><span class="kw1148">std::<span class="me2">strcmp</span></span></a><span class="br0">(</span>what<span class="br0">(</span><span class="br0">)</span>, other.<span class="me1">what</span><span class="br0">(</span><span class="br0">)</span><span class="br0">)</span> <span class="sy1">==</span> <span class="nu0">0</span></code>.</span> <span><span class="t-mark-rev t-since-cxx11">(since C++11)</span></span></span>
</div> <h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> what_arg </td> <td> - </td> <td> explanatory string </td>
</tr> <tr class="t-par"> <td> other </td> <td> - </td> <td> another exception object to copy </td>
</tr>
</table> <h3 id="Exceptions"> Exceptions</h3> <div class="t-li1">
<span class="t-li">1-2)</span> May throw <code><a href="../memory/new/bad_alloc" title="cpp/memory/new/bad alloc">std::bad_alloc</a></code>
</div> <h3 id="Notes"> Notes</h3> <p>Because copying <code>std::domain_error</code> is not permitted to throw exceptions, this message is typically stored internally as a separately-allocated reference-counted string. This is also why there is no constructor taking <code>std::string&amp;&amp;</code>: it would have to copy the content anyway.</p>
</div> <div class="t-member"> <h2 id="std::domain_error::operator.3D">  std::domain_error::operator= </h2> <table class="t-dcl-begin">   <tr class="t-dcl t-until-cxx11"> <td> <pre data-language="cpp">domain_error&amp; operator=( const domain_error&amp; other );</pre>
</td> <td class="t-dcl-nopad"> </td> <td> <span class="t-mark-rev t-until-cxx11">(until C++11)</span> </td> </tr> <tr class="t-dcl t-since-cxx11"> <td> <pre data-language="cpp">domain_error&amp; operator=( const domain_error&amp; other ) noexcept;</pre>
</td> <td class="t-dcl-nopad"> </td> <td> <span class="t-mark-rev t-since-cxx11">(since C++11)</span> </td> </tr>   </table> <p>Assigns the contents with those of <code>other</code>. <span class="t-rev-inl t-since-cxx11"><span>If <code>*this</code> and <code>other</code> both have dynamic type <code>std::domain_error</code> then <code><a href="http://en.cppreference.com/w/cpp/string/byte/strcmp"><span class="kw1148">std::<span class="me2">strcmp</span></span></a><span class="br0">(</span>what<span class="br0">(</span><span class="br0">)</span>, other.<span class="me1">what</span><span class="br0">(</span><span class="br0">)</span><span class="br0">)</span> <span class="sy1">==</span> <span class="nu0">0</span></code> after assignment.</span> <span><span class="t-mark-rev t-since-cxx11">(since C++11)</span></span></span>.</p>
<h3 id="Parameters_2"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> other </td> <td> - </td> <td> another exception object to assign with </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <p><code>*this</code>.</p>
</div> <div class="t-member"> <h2 id="std::domain_error::what">  std::domain_error::what </h2> <table class="t-dcl-begin">   <tr class="t-dcl t-until-cxx11"> <td> <pre data-language="cpp">virtual const char* what() const throw();</pre>
</td> <td class="t-dcl-nopad"> </td> <td> <span class="t-mark-rev t-until-cxx11">(until C++11)</span> </td> </tr> <tr class="t-dcl t-since-cxx11"> <td> <pre data-language="cpp">virtual const char* what() const noexcept;</pre>
</td> <td class="t-dcl-nopad"> </td> <td> <span class="t-mark-rev t-since-cxx11">(since C++11)</span> </td> </tr>   </table> <p>Returns the explanatory string.</p>
<h3 id="Parameters_3"> Parameters</h3> <p>(none).</p>
<h3 id="Return_value_2"> Return value</h3> <p>Pointer to a null-terminated string with explanatory information. The string is suitable for conversion and display as a <code><a href="../string/basic_string" title="cpp/string/basic string">std::wstring</a></code>. The pointer is guaranteed to be valid at least until the exception object from which it is obtained is destroyed, or until a non-const member function (e.g. copy assignment operator) on the exception object is called.</p>
<h3 id="Notes_2"> Notes</h3> <p>Implementations are allowed but not required to override <code>what()</code>.</p>
</div> <div class="t-inherited"> <h2 id="Inherited_from_std::logic_error"> Inherited from <a href="logic_error" title="cpp/error/logic error">std::logic_error</a> </h2> </div> <div class="t-inherited"> <h2 id="Inherited_from_std::exception"> Inherited from <a href="exception" title="cpp/error/exception"> std::exception</a> </h2> <h3 id="Member_functions_2"> Member functions</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="exception/~exception" title="cpp/error/exception/~exception"> <span class="t-lines"><span>(destructor)</span></span></a></div>
<div><span class="t-lines"><span><span class="t-cmark">[virtual]</span></span></span></div> </td> <td> destroys the exception object <br> <span class="t-mark">(virtual public member function of <code>std::exception</code>)</span>  </td>
</tr> <tr class="t-dsc"> <td> <div><a href="exception/what" title="cpp/error/exception/what"> <span class="t-lines"><span>what</span></span></a></div>
<div><span class="t-lines"><span><span class="t-cmark">[virtual]</span></span></span></div> </td> <td> returns an explanatory string <br> <span class="t-mark">(virtual public member function of <code>std::exception</code>)</span>  </td>
</tr> </table> </div>           <div class="_attribution">
  <p class="_attribution-p">
    &copy; cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/error/domain_error" class="_attribution-link">https://en.cppreference.com/w/cpp/error/domain_error</a>
  </p>
</div>
